File: packages/sdk-python/agenttv/context.py
"""Async AgentTV context — thin wrapper for emitting events to the sidecar."""

from __future__ import annotations

import os
import time
from typing import Any

import httpx


class AgentTVContext:
    """Thin wrapper — under 100 lines — with zero framework dependencies.

    Sends events to the local sidecar process via HTTP POST.
    The sidecar handles validation, forwarding, and budget enforcement.
    """

    def __init__(self) -> None:
        self._base_url = os.environ.get("AGENTTV_SIDECAR_URL", "http://localhost:8080")
        self._start_time = time.time()
        self._client: httpx.AsyncClient | None = None

    @property
    def elapsed(self) -> float:
        """Seconds since agent started."""
        return time.time() - self._start_time

    async def _get_client(self) -> httpx.AsyncClient:
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(base_url=self._base_url, timeout=5.0)
        return self._client

    async def emit_action(self, name: str, data: dict[str, Any]) -> None:
        """Emit an action event. Required: data['description']."""
        if "description" not in data:
            raise ValueError("action events require 'description' in data")
        await self._emit("action", name, data)

    async def emit_revenue(self, name: str, data: dict[str, Any]) -> None:
        """Emit a revenue event. Required: data['amount'], data['currency']."""
        if "amount" not in data or "currency" not in data:
            raise ValueError("revenue events require 'amount' and 'currency' in data")
        await self._emit("revenue", name, data)

    async def emit_status(self, data: dict[str, Any]) -> None:
        """Emit a status heartbeat. Required: data['state'], data['uptime'], data['budget_left']."""
        for key in ("state", "uptime", "budget_left"):
            if key not in data:
                raise ValueError(f"status events require '{key}' in data")
        await self._emit("status", "heartbeat", data)

    async def emit_error(self, data: dict[str, Any]) -> None:
        """Emit an error event. Required: data['message'], data['severity']."""
        if "message" not in data or "severity" not in data:
            raise ValueError("error events require 'message' and 'severity' in data")
        if data["severity"] not in ("warning", "critical", "fatal"):
            raise ValueError("severity must be 'warning', 'critical', or 'fatal'")
        await self._emit("error", "error", data)

    async def _emit(self, event_type: str, event_name: str, data: dict[str, Any]) -> None:
        """POST to sidecar."""
        client = await self._get_client()
        payload = {
            "event_type": event_type,
            "event_name": event_name,
            "data": data,
        }
        try:
            resp = await client.post("/emit", json=payload)
            resp.raise_for_status()
        except httpx.HTTPStatusError as exc:
            # 403 means budget exceeded — let the caller know
            if exc.response.status_code == 403:
                raise RuntimeError("AgentTV budget exceeded — agent will be terminated") from exc
            raise
        except httpx.ConnectError:
            # Sidecar not running — silently drop (agent shouldn't crash)
            pass

    async def close(self) -> None:
        """Close the underlying HTTP client."""
        if self._client and not self._client.is_closed:
            await self._client.aclose()

File: packages/sdk-node/src/context.ts
/**
 * AgentTV SDK context — emits events to the local sidecar process.
 *
 * Zero dependencies: uses Node.js built-in fetch (available since Node 18).
 */

// ---------------------------------------------------------------------------
// Event type interfaces
// ---------------------------------------------------------------------------

export interface ActionData {
  description: string;
  [key: string]: unknown;
}

export interface RevenueData {
  amount: number;
  currency: string;
  [key: string]: unknown;
}

export interface StatusData {
  state: string;
  uptime: number;
  budget_left: number;
  [key: string]: unknown;
}

export interface ErrorData {
  message: string;
  severity: "warning" | "critical" | "fatal";
  [key: string]: unknown;
}

export interface AgentEvent {
  event_type: "action" | "revenue" | "status" | "error";
  event_name: string;
  data: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Context
// ---------------------------------------------------------------------------

export class AgentTVContext {
  private baseUrl: string;
  private startTime: number;

  constructor() {
    this.baseUrl = process.env.AGENTTV_SIDECAR_URL || "http://localhost:8080";
    this.startTime = Date.now();
  }

  /** Seconds since the context was created. */
  get elapsed(): number {
    return (Date.now() - this.startTime) / 1000;
  }

  /**
   * Emit an action event.
   * @param name  - A short identifier for the action (e.g. "search", "click").
   * @param data  - Must include `description`.
   */
  async emitAction(name: string, data: ActionData): Promise<void> {
    if (!data.description) {
      throw new Error("action events require data.description");
    }
    await this._emit("action", name, data);
  }

  /**
   * Emit a revenue event.
   * @param name  - A short identifier (e.g. "booking", "sale").
   * @param data  - Must include `amount` (number) and `currency` (string).
   */
  async emitRevenue(name: string, data: RevenueData): Promise<void> {
    if (typeof data.amount !== "number") throw new Error("revenue events require data.amount (number)");
    if (typeof data.currency !== "string") throw new Error("revenue events require data.currency (string)");
    await this._emit("revenue", name, data);
  }

  /**
   * Emit a status heartbeat.
   * @param data - Must include `state`, `uptime`, and `budget_left`.
   */
  async emitStatus(data: StatusData): Promise<void> {
    if (typeof data.state !== "string") throw new Error("status events require data.state (string)");
    if (typeof data.uptime !== "number") throw new Error("status events require data.uptime (number)");
    if (typeof data.budget_left !== "number") throw new Error("status events require data.budget_left (number)");
    await this._emit("status", "heartbeat", data);
  }

  /**
   * Emit an error event.
   * @param data - Must include `message` and `severity` (warning | critical | fatal).
   */
  async emitError(data: ErrorData): Promise<void> {
    if (!data.message) throw new Error("error events require data.message");
    if (!["warning", "critical", "fatal"].includes(data.severity)) {
      throw new Error("error events require data.severity (warning | critical | fatal)");
    }
    await this._emit("error", "error", data);
  }

  /** POST an event to the sidecar. */
  private async _emit(eventType: string, eventName: string, data: Record<string, unknown>): Promise<void> {
    const payload: AgentEvent = {
      event_type: eventType as AgentEvent["event_type"],
      event_name: eventName,
      data,
    };

    try {
      const res = await fetch(`${this.baseUrl}/emit`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });

      if (res.status === 403) {
        throw new Error("AgentTV budget exceeded — agent will be terminated");
      }

      if (!res.ok) {
        const body = await res.text();
        console.error(`[agenttv] Sidecar responded ${res.status}: ${body}`);
      }
    } catch (err) {
      // Connection refused = sidecar not running. Don't crash the agent.
      if (err instanceof TypeError && (err.message.includes("fetch") || err.message.includes("ECONNREFUSED"))) {
        return;
      }
      throw err;
    }
  }
}
